ArduPlane/adsb.cpp

```diff
@@ -32,70 +32,95 @@
-            break;
-        }
-        switch(behavior) {
-        case AP_ADSB::ADSB_BEHAVIOR_NONE:
-        default:
-            break;
+    if (control_mode != AUTO) {
+        // evasion is only supported while in AUTO mode
+        return false;
+    }
+    if (mission.get_current_nav_cmd().id == MAV_CMD_NAV_TAKEOFF ||
+        mission.get_current_nav_cmd().id == MAV_CMD_NAV_LAND) {
+        // evasion during a plane take-off or land can result in a crash. Ignore ADS-B traffic
+        return false;
+    }
 
-        case AP_ADSB::ADSB_BEHAVIOR_LOITER:
-        case AP_ADSB::ADSB_BEHAVIOR_LOITER_AND_DESCEND:
-            if (adsb.get_another_vehicle_within_radius() && !adsb.get_is_evading_threat()) {
-                adsb.set_is_evading_threat(true);
-                gcs_send_text(MAV_SEVERITY_CRITICAL, "ADS-B threat found, performing LOITER");
-                adsb_state.prev_wp = prev_WP_loc;
-                set_mode(LOITER);
-                if (behavior == AP_ADSB::ADSB_BEHAVIOR_LOITER_AND_DESCEND) {
-                    adsb_state.time_last_alt_change_ms = now;
-                }
-            }
-        } // switch behavior
-        break; // case auto
+    adsb_state.timestamp_ms =  millis();
+    adsb_state.prev_wp = prev_WP_loc; // must be done BEFORE set_mode()
 
-    case LOITER:
-        switch(behavior) {
-        case AP_ADSB::ADSB_BEHAVIOR_NONE:
-            // TODO: recover from this
-        default:
-            break;
+    switch(adsb.get_behavior()) {
+    case AP_ADSB::ADSB_BEHAVIOR_NONE:
+    default:
+        gcs_send_text(MAV_SEVERITY_CRITICAL, "ADS-B threat found, no action taken");
+        break;
 
-        case AP_ADSB::ADSB_BEHAVIOR_LOITER:
-        case AP_ADSB::ADSB_BEHAVIOR_LOITER_AND_DESCEND:
-            if (adsb.get_is_evading_threat()) {
-                if (!adsb.get_another_vehicle_within_radius()) {
-                    adsb.set_is_evading_threat(false);
-                    gcs_send_text(MAV_SEVERITY_CRITICAL, "ADS-B threat gone, continuing mission");
-                    set_mode(AUTO);
-                    prev_WP_loc = adsb_state.prev_wp;
-                    auto_state.no_crosstrack = false;
-                } else if (behavior == AP_ADSB::ADSB_BEHAVIOR_LOITER_AND_DESCEND &&
-                            now - adsb_state.time_last_alt_change_ms >= 1000) {
-                   // slowly reduce altitude 1m/s while loitering. Drive into the ground if threat persists
-                    adsb_state.time_last_alt_change_ms = now;
-                    next_WP_loc.alt -= 100;
-                } // get_another_vehicle_within_radius
-            } // get_is_evading_threat
-        } // switch behavior
-        break; // case LOITER
+    case AP_ADSB::ADSB_BEHAVIOR_GUIDED:
+        gcs_send_text(MAV_SEVERITY_CRITICAL, "ADS-B threat found, switching to GUIDED mode");
+        set_mode(GUIDED);

```

Comment:
Sorry, I didn't remembered that set_mode returned void in Plane. I was thinking in the case that GPS is lost that you couldn't go into GUIDED or LOITER but I'm probably confusing it with Copter.